Remove login/logout components and redirect to urls

tdeekens opened the pull request:
#### Summary

This pull requests removes the login/logout components from the app-kit while redirecting to the urls.

emmenko commented on packages/application-shell/src/components/application-shell/application-shell.js:
```diff
@@ -438,17 +429,19 @@ export default class ApplicationShell extends React.Component {
+                   *   Given the application is in development mode (locally) we redirect to the application on staging.
+                   */
+                  const frontendHost =
+                    process.env.NODE_ENV === 'production'
```
We should use `servedByProxy` to determine if we're running locally or on a production environment

emmenko commented on packages/application-shell/src/components/application-shell/application-shell.js:
```diff
@@ -438,17 +429,19 @@ export default class ApplicationShell extends React.Component {
+                   */
+                  const frontendHost =
+                    process.env.NODE_ENV === 'production'
+                      ? ''
```
Should we use `window.location.origin`? I find it clearer on the intent

tdeekens replied:
Yeah makes sense. My only goal was to have it working on all cloud deployments.

emmenko commented on packages/application-shell/src/components/application-shell/application-shell.js:
```diff
@@ -438,17 +429,19 @@ export default class ApplicationShell extends React.Component {
+                  const frontendHost =
+                    process.env.NODE_ENV === 'production'
+                      ? ''
+                      : this.props.environment.frontendHost;
```
I don't think we can use this. For local development this is `localhost:3001`. We should probably hardcode the staging domain here, or we add a new value e.g. `loginUrl` or `authUrl` in `env.json`

tdeekens replied:
Can you elaborate why `frontendHost` is not working? Adding `loginUrl` sounds reasonable but I don't follow how that is not just `${frontendHost}/login`.

emmenko replied:
Well, in development `frontendHost` points to `localhost:3001`. So when you go to `/login`, we end up in an infinite loop because the page keeps reloading as there is no "login service" running.

Instead, it should point to `https://mc.commercetools.com/login`, right? (or our staging domain, which I won't write down).

tdeekens replied:
Understood. Thanks. Then I will add a new env variable called `authUrl` to have it be "augmented" to `/login` and `/logout`.

emmenko commented on packages/application-shell/src/index.js:
```diff
@@ -18,6 +18,10 @@ export {
 export { default as AsyncChunkLoader } from './components/async-chunk-loader';
 export { GtmContext } from './components/gtm-booter';
+export {
+  default as GtmUserLogoutTracker,
```
Should we move this to the auth app? It's not used anymore here.

tdeekens replied:
I thought about it but then we spread logic: set-up in app-kit and teardown in app. That's why I liked to just export it.

emmenko replied:
Yeah I guess...

emmenko commented on packages/application-shell/src/components/application-shell-provider/application-shell-provider.js:
```diff
@@ -61,47 +60,32 @@ export default class ApplicationShellProvider extends React.Component {
-                     * TODO: do a hard page reload once we move the login routes
-                     * to a different app.
-                     */}
-                    <Route path="/logout" component={Logout} />
```
This needs to stay and to do a hard redirect. Unless we change the `logout` link to be a hard redirect

tdeekens replied:
I added a hard-redirect in the same vein as in the other component now.

tdeekens commented:
@emmenko pushed a commit containing some fixes.

emmenko commented:
Ok thanks. Have you seen this comment btw? (https://github.com/commercetools/merchant-center-application-kit/pull/444#discussion_r264612202)

I don't think we can use `frontendHost` value.

tdeekens commented on playground/env.prod-us.json:
```diff
@@ -1,5 +1,6 @@
 {
   "frontendHost": "mc.commercetools.co",
+  "mcAuthUrl": "https://mc.commercetools.co",
```
Thanks. I wanted to do this after I saw that we have an `ctpAuthUrl` as an env variable.

tdeekens commented on packages/application-shell/src/components/application-shell-provider/application-shell-provider.js:
```diff
@@ -18,6 +18,26 @@ import ErrorApologizer from '../error-apologizer';
+    return null;
+  }
+}
+
```
Many thanks!

emmenko commented on packages/application-shell/src/components/login/login.js:
```diff
@@ -1,257 +0,0 @@
-        // Inspect the target url to see if the host is one of commercetools
-        try {
-          const nextTargetUrlObject = new URL(nextTargetUrl);
-          if (nextTargetUrlObject.hostname === window.location.hostname) {
```
This is currently preventing the redirect from staging -> localhost (currently a problem only for development).

tdeekens replied:
We would change that in the mc-app now though, right? Paring it somehow with `servedByProxy` again I suppose...

emmenko replied:
I guess yes

tdeekens replied:
We can tackle that on the PR taking a pre-canary build of this I guess.

tdeekens commented on packages/application-shell/src/components/authenticated/authenticated.js:
```diff
@@ -1,17 +1,54 @@
+  const cachedAuthenticationState = hasCachedAuthenticationState();
+  if (cachedAuthenticationState) {
+    return props.render({ isAuthenticated: true });
+  }
```
Still any subsequent request will fail in the mc-be in turn logging the user out again. So there is no harm in this caching here, correct?

emmenko replied:
Yes correct

tdeekens commented on packages/application-shell/src/components/authenticated/authenticated.js:
```diff
@@ -1,17 +1,54 @@
+        // Even though the login page might set this flag, we just make sure that
+        // we do it here as well. This will help in the future when we eventually
+        // move the auth service to itw own domain, in which case the local storage
+        // is not shared anymore.
```
You're saying that in that case e.g. `login.commercetools.com` might save things to local storage as well as this but as code has/may run on different domains it remains to work?

emmenko replied:
Yes that's the idea, to probably reduce breaking changes.
In case we have our own login/auth service, I would still like to consider a different auth mechanism though (as we discussed already).

tdeekens replied:
Yeah.

tdeekens commented on packages/application-shell/src/components/authenticated/authenticated.js:
```diff
@@ -1,17 +1,54 @@
+    >
+      {({ loading, data, error }) => {
+        if (error) {
+          console.error(error);
```
Dangling log?

emmenko replied:
Not really, I wasn't sure what to do with that to be honest. Most likely is going to be a "unauthorized" error, which we can safely ignore.

tdeekens replied:
Should we maybe report to sentry instead?

emmenko replied:
We will get spammed if we log on every unauthorized. Maybe we can send errors that are not unauthorized.

tdeekens commented on packages/application-shell/src/components/authenticated/authenticated.js:
```diff
@@ -1,17 +1,54 @@
+        if (!loading && data && data.amILoggedIn) {
+          return props.render({ isAuthenticated: true });
+        }
+        return null;
```
Q: wny not return `isAuthenticated: false` in this case?

emmenko replied:
If we do that, we end up rendering e.g. the login page. However, we need to wait for the request to be finished, and only then render the function (that's also why I renamed it from `children` to `render`, as it's only executed in certain cases)

tdeekens replied:
Ah, it is too late. Missed that it is a render-faac not a children-faac. Makes sense. Thanks.

tdeekens commented on packages/application-shell/src/components/application-shell-provider/application-shell-provider.spec.js:
```diff
@@ -32,6 +31,10 @@ const createTestProps = props => ({
 describe('rendering', () => {
+  beforeEach(() => {
+    // eslint-disable-next-line global-require
+    require('../authenticated').setIsAuthenticated('true');
```
Q: Can't we just import this as `import { setIsAuthenticated } from '../authenticated'`? I presume, from reading the code, you ran into issues with mocking in L11. I find this `require`-cjs work around a bit confusing. Would a 

```js
jest.mock('../authenticated', () => ({
  ...require.requireActual('../authenticated'),
  // eslint-disable-next-line react/prop-types
  Authenticated: ({ children }) => <div>{children}</div>,
}));
```

Might be a bit trouble some with the default export of `Authenticated` and I am not 100% sure what the issue here is but just wanted to share the thouhgt.

emmenko replied:
I've spent a bit of time on this actually.

The main problem is that the `ApplicationShellProvider` renders the normal apollo provider. So if we test it and wrap it with an apollo mock provider to mock the graphql request, it won't work because the normal apollo provider takes precedence over the mocked one. This is only a "problem" with this component, as it's the top level one.

So the only solution is to mock the `Authenticated` component using a manual mock, and we need to be able to control the `isAuthenticated` value from outside of the mock.

I haven't tried with `import { setIsAuthenticated } from '../authenticated'`, will do that tomorrow. Thanks!

tdeekens commented:
A bit strange that the e2e tests for the playground keep failing. 😩😭

emmenko commented:
It’s because of the login redirect I believe. Don’t worry too much about the e2e tests for now.

tdeekens commented:
BTW: I published an earlier version to pre-canary. Will publish this there later too so we can try it.

![CleanShot 2019-03-13 at 11 22 01@2x](https://user-images.githubusercontent.com/1877073/54271657-40fe8d80-4582-11e9-8a90- 262b6fa840bc.png)

emmenko commented:
New release:

```
Successfully published:
 - @commercetools-frontend/application-shell@11.0.3-pre-canary.11+f3b22193
```

emmenko commented:
Update: for local development we need to use a login page from `localhost`, as the API response needs to set the cookie. Using a login page to another domain doesn't work in that regards.
Therefore, we don't need the `mcAuthUrl` anymore.

To amend that (for the short period), we serve a "simple" html login page from the webpack dev server, that logs the user in and redirects to the actual app running on `localhost`.

@tdeekens please try this locally with the playground app to see if everything works fine, thanks!

tdeekens commented on packages/mc-scripts/config/views/login.pug:
```diff
@@ -0,0 +1,40 @@
+html
```
Nice. Sorry to ask but do we really need pug for this? 😄

emmenko replied:
We don't, but on the other hand...why not? 😄

We can use plain html, I don't mind...as long as it works

tdeekens commented:
> @tdeekens please try this locally with the playground app to see if everything works fine, thanks!

Sure. I planned to have a little intro meeting next week with all FE engineers going through JWTs and how login/sign-up will work. There I planned to also elaborate on why we need this page for now.

tdeekens commented:
Confirmed sir. The temporary development login page works like a charm. Could use some more Netscape style styling but is lovely.

emmenko commented:
New release:

```
Successfully published:
 - @commercetools-frontend/application-shell@11.0.3-pre-canary.13+907f42f7
 - @commercetools-frontend/mc-scripts@11.0.3-pre-canary.13+907f42f7
```

tdeekens commented:
👑 Majestic 👑

tdeekens commented on packages/mc-scripts/config/views/login.js:
```diff
@@ -20,19 +19,25 @@ window.addEventListener('load', function loaded() {
-          window.location.replace(redirectTo);
-        } else {
+    window
+      .fetch(data.get('url'), {
```
Nice. Wanted to ask why we not using fetch after thinking about this.

tdeekens commented on packages/mc-scripts/config/views/logout.pug:
```diff
@@ -1,4 +1,10 @@
+  body
+    div
+      h3 This is the logout page for local development.
+      p Be aware that you might still have an active session as the cookie is assigned to a production domain (e.g. commercetools.com) which we can't unset from localhost. This is only a problem on local development and we will fix this in the future.
```
Very cool! Thanks!

tdeekens commented on packages/application-shell/src/test-utils/test-utils.js:
```diff
@@ -285,6 +292,15 @@ const experimentalRender = (ui, renderOptions) => {
+    environment: defaultEnvironment,
+    permissions: defaultPermissions,
+  },
+};
```
Just an idea. It makes the tests a bit simpler. We could also expose the default rtl render there. /cc @emmenko

emmenko replied:
> Just an idea. It makes the tests a bit simpler.

What do you mean? Those values are already available as

```js
const { user, project, environment } = renderApp(<Foo />);
```

> We could also expose the default rtl render there

What do you mean? It's already exported with `export * from 'react-testing-library';`

tdeekens replied:
I know. It would just be easier to to test the test-utils. Anyway. Will remove.

emmenko commented on playground/src/components/state-machines-list/state-machines-list.spec.js:
```diff
@@ -25,16 +25,22 @@ const createStateMachinesSdkMock = () => ({
   },
 });
 
+const renderApp = (options = {}) => {
```
Careful because `renderApp` is exported from `test-utils`

tdeekens replied:
Not on this branch yet. After I rebase, yes.

tdeekens commented:
NOTE: rebased to have CircleCI take over testing here.

tdeekens commented:
Huuuuray. The branch is green. Given we squash merge anyway I won't clean up the commit history here.

I think generally it's ready for review though.

tdeekens commented:
This branch can only be merged after UAT has passed on the respective internal branch deployment.

emmenko commented on packages/application-shell/src/components/project-container/project-container.js:
```diff
@@ -112,9 +121,19 @@ export class ProjectContainer extends React.Component {
+     * is requested to render a full page redirect (to have the proxy serve the request) occurs.
+     */
     if (this.props.user && this.props.user.projects.total === 0)
-      return <Redirect to="/logout?reason=no-projects" />;
```
Q: what should we do here as long as signup is disabled?
Should we wrap this logic into feature toggle and clean up afterwards?

tdeekens replied:
Hm. Is there a problem with not wrapping it? In the AC you get your project and org straight away. If you get an account otherwise it may be nice to get to account creation (another toggle). Do you see a specific problem we can run into?

emmenko replied:
Just thinking about the case when the user does not have any projects (it’s not necessarily after signup) and we redirect to create a project but the accounts app is non available (feature toggled).

tdeekens replied:
e46235f

emmenko commented on packages/i18n/data/core.json:
```diff
@@ -36,30 +36,6 @@
-  "LoginLocked.paragraph1": "For security reasons, your account has been locked because of too many failed log-in attempts. You can {resetPasswordLink} or contact the commercetools {helpDeskLink}",
-  "LoginLocked.resetPasswordLink": "reset your password",
-  "LoginLocked.title": "Account Locked",
-  "LoginPage.privacyPolicy": "Privacy Policy",
```
Let’s also remove the translations for this

tdeekens replied:
2168293

emmenko commented on playground/src/components/entry-point/entry-point.js:
```diff
@@ -28,6 +28,11 @@ export const ApplicationStateMachines = () => (
     <Route path="/:projectKey/state-machines" component={AsyncStateMachines} />
+    {/* Given a user does not have any projects yet, then the app-shell redirects here. The playground renders a message instead of letting the <RouteCatchAll> handle it. */}
+    <Route
+      path="/account/projects/new"
```
Hmm how is this useful? There is no account all running locally

tdeekens replied:
If I remember correctly otherwise you and up in a redirect loop as the route catch all kicks in and so forth. Plus it felt kind of okay to have it for debugging as there is no accounts app.

emmenko replied:
Hmm I see. Does it mean that every app needs to do that?

tdeekens replied:
Fair question. Something we should check, true. Maybe we should redirect only if it's served by the proxy of something smater.

tdeekens replied:
See below.

emmenko commented on packages/application-shell/src/components/project-container/project-container.js:
```diff
@@ -28,6 +28,15 @@ const shouldShowNotificationForTrialExpired = daysLeft =>
   daysLeft <= minDaysToDisplayNotification &&
   daysLeft >= maxDaysToDisplayNotification;
 
+export class RedirectToProjectCreate extends React.Component {
```
I think we can fix that problem here:
- if it’s not `servedByProxy` we don’t redirect and show a simple message

If we’re in the account app already, there won’t be any redirect and the route will match and let the app handle the rest.

tdeekens replied:
Yup I agree. Thanks. Have a good weekend 🤪

tdeekens replied:
843ae4c

emmenko commented:
@tdeekens should I take care of the rebase of you do that?

tdeekens commented:
I already did. Thanks for asking.

emmenko commented:
Ok, because there are a bunch of conflicts here

emmenko commented:
> why did lint-staged not catch this

Yeah, I think because you don't create any commits during rebase. Right?

emmenko commented:
We also need to do a new pre-canary release now, if you want the tests to pass

tdeekens commented:
> We also need to do a new pre-canary release now, if you want the tests to pass

Yeap I was just making ☕️ and about to do that now :)

emmenko commented on application-templates/starter/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "merchant-center-application-template-starter",
-  "version": "12.2.0",
+  "version": "12.0.1-pre-canary.32",
```
Q: should we revert those version before merging? I guess it shouldn't matter if we then release a major, all version will be bumped anyway.

tdeekens replied:
What do you mean by revert? Unpublishing won't work after 24 hours. We could remove the dist-tag.

emmenko commented on packages/application-shell/src/components/project-container/project-container.js:
```diff
@@ -112,9 +124,32 @@ export class ProjectContainer extends React.Component {
+      return (
+        <Switch>
+          <Route path="/account" render={this.props.render} />
+          {this.props.environment.servedByProxy && (
```
Hmm that's not what I had in mind. We should still render the `RedirectToProjectCreate` but instead of rendering `null`, we should render a message explaining that there are no projects and that you would need the account app blabla. And of course the redirect should be prevented.

We can also have another component and render one or the other. But the point is to keep the route, otherwise things will probably break.

tdeekens replied:
K

tdeekens commented:
@jonnybel can you also have a look when you have time please? We can't merge it yet but it would be cool to have it ready.

jonnybel commented:
Yes, I've looked at parts of it while it's being worked on, gotta take a final look.

emmenko commented:
@tdeekens is there an ETA on when we are supposed to merge this?

tdeekens commented:
Alright. Rebased out all pre-canary releases. So the pkg.json's are now up-to-date with master. Furthermore in 8e95ddd I added an `aria-label` for sign up cause I realised otherwise running the mc locally will not work the the mc-fe E2E tests as it requires aria-label.
